File: 0020-valid-parentheses/0020-valid-parentheses.java
import java.util.*;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<>();
        Map<Character, Character> parse = new HashMap<>() {{
            put(')', '(');
            put(']', '[');
            put('}', '{');
        }};
        
        for(char ch: s.toCharArray()) {
            if(ch == '(' || ch == '[' || ch == '{') st.push(ch);
            else {
                if(st.empty() || st.peek() != parse.get(ch)) return false;
                st.pop();
            }
        }
        
        return st.empty();
    }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        st = []
        paren = {
            ')':'(',
            '}':'{',
            ']':'['
        }
        
        for ch in s:
            if ch not in paren:
                st.append(ch)
            elif not(st and paren[ch] == st.pop()):
                return False
        
        return len(st) == 0